diff --git a/node_modules/react-native-record-screen/ios/RecordScreen.mm b/node_modules/react-native-record-screen/ios/RecordScreen.mm
index 67ac51f..157f07c 100644
--- a/node_modules/react-native-record-screen/ios/RecordScreen.mm
+++ b/node_modules/react-native-record-screen/ios/RecordScreen.mm
@@ -290,12 +290,5 @@ - (int) adjustMultipleOf2:(int)value;
 }
 
 // Don't compile this code when we build for the old architecture.
-#ifdef RCT_NEW_ARCH_ENABLED
-- (std::shared_ptr<facebook::react::TurboModule>)getTurboModule:
-    (const facebook::react::ObjCTurboModule::InitParams &)params
-{
-    return std::make_shared<facebook::react::NativeRecordScreenSpecJSI>(params);
-}
-#endif
 
 @end
